File: leetcode/java/0015. 3Sum.java
package leetcode.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// Time complexity: O(N^2)
// Space complexity: O(N)

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i=0; i<nums.length; i++) {
            int target = -nums[i];

            int left = i+1;
            int right = nums.length - 1;

            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }

            while (right > left) {
                if (nums[right] + nums[left] > target) {
                    right--;
                } else if (nums[right] + nums[left] < target) {
                    left++;
                } else {
                    res.add(List.of(nums[i], nums[left], nums[right]));
                    left++;
                    while (right > left && nums[left] == nums[left-1]) {
                        left += 1;
                    }
                }
            }
        }
        return res;
    }
}

File: leetcode/python/0015. 3Sum.py
# Time complexity: O(N^2)
# Space complexity: O(N)

class Solution(object):
    def threeSum(self, nums):
        res = []
        nums.sort()

        for idx,num in enumerate(nums):
            target = -num

            left = idx+1
            right = len(nums) - 1

            if nums[idx] == nums[idx-1] and idx>0:
                continue

            while right > left:
                if nums[left] + nums[right] > target:
                    right -= 1
                elif nums[left] + nums[right] < target:
                    left += 1
                else:
                    res.append([num, nums[left], nums[right]])
                    left += 1
                    while right > left and nums[left] == nums[left-1]:
                        left += 1
          
        return res    